### DISCUSSION TITLE
How to test a visitor that uses visit_mut_script

### DISCUSSION THREAD

**Discussion by leafypout** [swc-project/swc#8712] [Q&A] [answered]
Let's say you have a visitor, something like this
```rust
struct Visitor;

impl VisitMut for Visitor {
    fn visit_mut_script(&mut self, script: &mut Script) {
        // Do something with script
    }
}
```

Then you create a unit test, something like
```rust
#[test]
fn test_visitor() {
    let input = "";
    let expected = "";

    test_transform(Syntax::default(), |_| as_folder(&mut Visitor), input, expected, true);

    // Optionally do some assertions with assert, assert_eq etc on the visitor
}
```

The test function won't work because the internal code for `test_transform` parses the input code as a module and not a script, so `visit_mut_script` is never called. Is there a different function to use instead of `test_transform` or is there a way to make `test_transform` use `parse_script` instead of `parse_module`?

**Comment by lubieowoce** [comment, marked as answer]
For anyone else stumbling upon this, `test_transform` takes an `is_module` argument now:
```rs
pub fn test_transform<F, P>(
    syntax: Syntax,
    is_module: Option<bool>,
    tr: F,
    input: &str,
    expected: &str,
)
where
    F: FnOnce(&mut Tester<'_>) -> P,
    P: Pass,
```
https://rustdoc.swc.rs/swc_ecma_transforms_testing/fn.test_transform.html

**Reply by leafypout** [reply]
Yep, I made a PR for this a while ago back when I first had this issue. `None` will make it use `parse_program`, `Some(true)` will make it use `parse_module` and `Some(false)` will make it use `parse_script`.

**Reply by leafypout** [reply]
Also given this change you should probably just make your visitor use `visit_mut_program` instead.
